nanocoat/lib/scritchui/scritchPencilRasterShapes.c

```diff
@@ -17,7 +17,399 @@
+	sjme_jboolean dot, dotFlip;
+	sjme_scritchui_line* clipLine;
+	sjme_scritchui_pencilDrawPixelFunc drawPixel;
+
+	
+	if (g == NULL)
+		return SJME_ERROR_NULL_ARGUMENTS;
+	
+	error = SJME_ERROR_NONE;
+
+	dot = SJME_JNI_TRUE;
+	dotFlip = (g->state.stroke == SJME_SCRITCHUI_PENCIL_STROKE_DOTTED);
+
+	drawPixel = g->prim.drawPixel;
+
+	/* Get clipping information. */
+	clipLine = &g->state.clipLine;
+	
+	/* Java's coordinate system has positive angles moving counter-clockwise */
+	arcAngle = -arcAngle;
+	startAngle = -startAngle;
+	
+
+	/**
+	 * This works similarly to Bresenham's midpoint circle algorithm. "steps" 
+	 * dictates how many iterations are used to draw the circle. A bigger value
+	 * will result in the same pixels being hit more times (and wasted cycles 
+	 * since they'll be discarded later) but will guarantee a perfectly filled 
+	 * outline, whereas a small value will result in gaps appearing in the 
+	 * circle since less points will be sampled. The current value is a good
+	 * balance between filling all positions on all kinds of shapes while 
+	 * hitting as few pixels as possible.
+	 */
+
+	centerX = x + w / 2.0f;
+	centerY = y + h / 2.0f;
+	radiusX = w / 2.0f;
+	radiusY = h / 2.0f;
+	startAngleRad = startAngle * 0.017453292f;
+	endAngleRad = ((startAngle + arcAngle) * 0.017453292f) 
+	- startAngleRad;
+	steps = fabs(arcAngle * ((w + h) / 2.0f) / 50.0f);
+	innerX, innerY;
+
+	firstFillX = lastFillX = firstFillY = lastFillY = -1;
+	
+	for (int i = 0; i < steps; i++) 
+	{
+		float angle = startAngleRad + (i * endAngleRad / steps);
+		
+		int innerX = round((centerX) + radiusX * cos(angle));
+		int innerY = round((centerY) + radiusY * sin(angle));
+		
+		/* Make sure we're not drawing out of bounds. */
+		if(innerX < clipLine->s.x || innerX >= clipLine->e.x
+			|| innerY < clipLine->s.y || innerY >= clipLine->e.y)
+			continue;
+
+		/* We cannot paint the same pixel more than once (breaks alpha) */
+		if((lastFillX == innerX && lastFillY == innerY) || 
+			(firstFillX == innerX && firstFillY == innerY)) 
+			continue;
+
+		if(i == 0) 
+		{ 
+			firstFillY = innerY; 
+			firstFillX = innerX; 
+		}
+		
+		lastFillX = innerX;
+		lastFillY = innerY;
+
+		/* If style is DOTTED, rendering will paint and skip pixels 1 by 1. */
+		if (dot)
+			error |= drawPixel(g, innerX, innerY);
+		dot ^= dotFlip;
+	}
+
+	/* Failed? */
+	if (sjme_error_is(error))
+		goto fail_any;
+		
+	/* Success? */
+	return error;
+	
+fail_any:
+	
+	return sjme_error_default(error);
+}
+
+sjme_errorCode sjme_scritchpen_corePrim_fillArc(
+	sjme_attrInNotNull sjme_scritchui_pencil g,
+	sjme_attrInValue sjme_jint x,
+	sjme_attrInValue sjme_jint y,
+	sjme_attrInPositive sjme_jint w,
+	sjme_attrInPositive sjme_jint h,
+	sjme_attrInValue sjme_jint startAngle,
+	sjme_attrInValue sjme_jint arcAngle)
+{
+	sjme_errorCode error;
+	sjme_scritchui_pencilDrawHorizFunc drawHoriz;
+	sjme_jint yz, yze, steps, innerX, innerY;
+	sjme_jboolean hasAlpha = 0;
```

Comment:
Initializer should be `SJME_JNI_FALSE`, it should also be later after everything as not all compilers support mixed initialization.